### PR TITLE
SE-2769: Migrate External Search Service to Payara

### PR DESCRIPTION
**Pull request by noah-dbc**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -20,11 +20,27 @@
     </properties>
```
Comment (mb-dbc-dk) [inline pom.xml]: parent from:

    <repositories>
        <repository>
            <id>mavenrepo.dbc.dk</id>
            <url>http://mavenrepo.dbc.dk/content/groups/public</url>
        </repository>
    </repositories>

    <parent>
        <groupId>dk.dbc</groupId>
        <artifactId>microservice-pom</artifactId>
        <version>latest-SNAPSHOT</version>
        <relativePath/>
    </parent>

File: pom.xml [hunk 2]
```diff
@@ -20,11 +20,27 @@
...
+                    <artifactId>maven-compiler-plugin</artifactId>
+                    <configuration>
+                        <source>1.8</source>
+                        <target>1.8</target>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+    </build>
+
```
Comment (mb-dbc-dk) [inline pom.xml]: Inherited from microservice-pom

File: service/src/main/java/dk/dbc/ess/service/ExternalSearchService.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+        this.maxPageSize = settings.getMaxPageSize();
+        this.executorService = Executors.newCachedThreadPool();
+        this.formatting = new Formatting(settings, metrics, client);
+        this.timerSruRequest = makeTimer(metrics, "sruRequest");
+        this.timerSruReadResponse = makeTimer(metrics, "sruReadResponse");
+        this.timerRequest = makeTimer(metrics, "Request");
+    }
+
+    @GET
+    @Path("rpn/")
```
Comment (mb-dbc-dk) [inline service/src/main/java/dk/dbc/ess/service/ExternalSearchService.java]: Trailing /

File: service/src/main/java/dk/dbc/ess/service/HowRU.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ * @author Noah Torp-Smith (nots@dbc.dk)
+ */
+@Path("howru")
+public class HowRU {
+    private final List<HealthCheck> checks;
+
+    public HowRU(HealthCheck... checks) {
+        this.checks = Arrays.asList(checks);
+    }
+
```
Comment (mb-dbc-dk) [inline service/src/main/java/dk/dbc/ess/service/HowRU.java]: Bean should be either: \@Stateless or \@Singleton

Pretty sure as is won't work as expected in payara
Will try default constructor to make an instance where howru can be called hence no health checks performed. (no checks implemented?)

File: service/src/main/java/dk/dbc/ess/service/ExternalSearchService.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+import org.w3c.dom.Element;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+
+/**
+ *
+ * @author Noah Torp-Smith (nots@dbc.dk)
+ */
+public class ExternalSearchService {
+    private static final Logger log = LoggerFactory.getLogger(ExternalSearchService.class);
```
Comment (mb-dbc-dk) [inline service/src/main/java/dk/dbc/ess/service/ExternalSearchService.java]: Should be \@Stateless To have working CDI

File: service/src/main/java/dk/dbc/ess/service/ExternalSearchService.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    Formatting formatting;
+
+    ExecutorService executorService;
+    Timer timerSruRequest;
+    Timer timerSruReadResponse;
+    Timer timerRequest;
+    private int maxPageSize;
+
+    public ExternalSearchService(Settings settings, MetricRegistry metrics) {
+        this.client = settings.getClientBuilder().build();
```
Comment (mb-dbc-dk) [inline service/src/main/java/dk/dbc/ess/service/ExternalSearchService.java]: Beans have (void) constructors. Configuration needs to be \@Inject'ed. If you need to initialize you can do it in a \@PostConstruct annotaed public void method.

File: pom.xml [hunk 7]
```diff
@@ -20,11 +20,27 @@
...
 </project>
```
Comment (thpdbc) [inline pom.xml]: Man kan muligvis blive fri fra Javadoc advarsler for genereret kode
(https://is.dbc.dk/job/ess-payara/job/SE-2769/8/analysis/)
ved at tilføje:
```
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <version>2.9.1</version>
                <artifactId>maven-javadoc-plugin</artifactId>
                <configuration>
                    <excludePackageNames>dk.dbc.openformat,dk.dbc.sru</excludePackageNames>
                </configuration>
            </plugin>
        </plugins>

```

File: service/src/main/java/dk/dbc/ess/service/ExternalSearchService.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+            rows = maxPageSize;
+        }
+        if (trackingId == null || trackingId.isEmpty()) {
+            trackingId = UUID.randomUUID().toString();
+        }
+        if (!knownBases.contains(base)) {
+            return serverError("Unknown base requested");
+        }
+        log.info("base: " + base + "; format: " + format +
+                "; start: " + start + "; rows: " + rows +
```
Comment (thpdbc) [inline service/src/main/java/dk/dbc/ess/service/ExternalSearchService.java]: Log framework ​slf4j
understøtter også formatering så man kan laves det med:
```
        log.info("base: {}; format: {}; start: {}; rows: {}; trackingId: {}; query: {}; type: {}",
                base, format, start, rows, trackingId, query, (isRPN ? "rpn" : "cql"));
```
Hvis man bruger det undergår man konkatenere strengen, hvis log er slået fra. Formateringen bliver ikke kaldt hvis loglevel er sat så skal printes

**Review by thpdbc — APPROVED** [review]

File: service/src/main/java/dk/dbc/ess/service/EssConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        if (clientBuilder == null) {
+            clientBuilder = ClientBuilder.newBuilder();
+        }
+        return clientBuilder;
+    }
+
+    protected Client getClient() {
+        if (client == null) {
+            client = getClientBuilder().build();
+        }
```
Comment (mb-dbc-dk) [inline service/src/main/java/dk/dbc/ess/service/EssConfiguration.java]: Put this in the \@PostConstruct

File: service/src/main/java/dk/dbc/ess/service/EssConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+            log.error("Exception: {}", ex.getMessage());
+        }
+        return new Properties();
+    }
+
+    protected ClientBuilder getClientBuilder() {
+        if (clientBuilder == null) {
+            clientBuilder = ClientBuilder.newBuilder();
+        }
+        return clientBuilder;
```
Comment (mb-dbc-dk) [inline service/src/main/java/dk/dbc/ess/service/EssConfiguration.java]: No need to store clientbuilder - only called once

**Review by mb-dbc-dk — CHANGES_REQUESTED** [review]
Lets take a look together

File: service/src/main/java/dk/dbc/ess/service/HowRU.java [hunk 11]
```diff
@@ -18,39 +18,72 @@
...
+     */
+    private boolean metaProxyUrlOk(String url) {
+        try {
+            URI uri = new URI(url);
+            Response r = essConfiguration.getClient().target(uri).request(MediaType.APPLICATION_XML_TYPE).get();
+            if (r.getStatus() == 200) {
+                return true;
+            }
+        } catch (Exception e) { }
+        return false;
```
Comment (mb-dbc-dk) [inline service/src/main/java/dk/dbc/ess/service/HowRU.java:74]: A little log about which service didn't respond will facilitate faster service debug

**Review by mb-dbc-dk — APPROVED** [review]
Just a little more log

File: pom.xml (hunk 7 above)
Comment (thpdbc) [reply pom.xml]: Kiggede lige igen på javadoc warnings for https://is.dbc.dk/job/ess-payara/job/master/21/

Konfigurations af maven-javadoc-plugin skal stå i projekt rod pom.xml, fordi det bliver kørt med "mvn javadoc:aggregate" i rod folderen

excludePackageNames skal være "dk.dbc.openformat,dk.dbc.sru.*", fordi openformat klassene ligger i dk.dbc.openformat pakken, men dk.dbc.sru klasserne ligger i subdirectories *

**Pull request merged**